eliot-service/eliot/cache.py

```diff
@@ -119,29 +177,16 @@ def set(self, key, data):
         This will log and emit metrics on OSError and IOError.
 
         :arg str key: the key to set
-        :arg bytes data: the data to save
+        :arg dict data: the data to save as a key/val dict
 
         """
-        result = ""
+        assert isinstance(data, dict)
+
         start_time = time.perf_counter()
         filepath = self.key_to_filepath(key)
 
-        # Save the file to a temp file and then rename that so as to avoid race
-        # conditions.
-        try:
-            temp_fp = tempfile.NamedTemporaryFile(
-                mode="w+b", suffix=".sym", dir=self.tmpdir, delete=False
-            )
-            temp_fp.write(data)
-            temp_fp.close()
-
-            filepath.parent.mkdir(parents=True, exist_ok=True)
-            Path(temp_fp.name).rename(filepath)
-            result = "success"
-
-        except OSError:
-            LOGGER.exception("Exception when writing to disk cache")
-            result = "fail"
+        ret = self.write_to_file(filepath, data)
+        result = "success" if ret else "fail"
```

Comment:
I kind of wish Markus grew a "structured metrics" system so I could store these tags in a context from wherever and pull them all out at once to generate the metric. Then I wouldn't have to do this awkward carrying through of values.